src/pybind/mgr/cephadm/serve.py

```diff
@@ -575,6 +570,33 @@ def _check_for_strays(self) -> None:
                 }
         self.mgr.set_health_checks(self.mgr.health_checks)
 
+    def _check_for_moved_osds(self) -> None:
+        all_osds: DefaultDict[int, List[orchestrator.DaemonDescription]] = defaultdict(list)
+        for dd in self.mgr.cache.get_daemons_by_type('osd'):
+            assert dd.daemon_id
+            all_osds[int(dd.daemon_id)].append(dd)
+        for osd_id, dds in all_osds.items():
+            if len(dds) <= 1:
+                continue
+            running = [dd for dd in dds if dd.status == DaemonDescriptionStatus.running]
+            error = [dd for dd in dds if dd.status == DaemonDescriptionStatus.error]
+            msg = f'Found duplicate OSDs: {", ".join(str(dd) for dd in dds)}'
+            logger.info(msg)
+            if len(running) != 1:
```

Comment:
why are we bailing out here if there isn't exactly one osd with this id running? Are was assuming the number running will only ever be 1 or 0?

Reply:
Those OSDs are gathered independently. Nothing prevents us from having two running osd daemons with the same ID. Maybe because the OSD was moved faster than our refresh interval (keep in mind we might have a 15 minute ssh timeout). Or the OSD is located on a host that is offline (we don't update the daemon running state, if the host is offline). Stuff like that.

Reply:
okay, so is the idea that we only know which osd to remove if only one of them is running so we can't do anything if that's not the case?

Reply:
better be safe than sorry is my approach here